sound/soc/sof/ipc4-topology.c

```diff
@@ -232,6 +232,38 @@ static int sof_ipc4_setup_process_init_byte_kcontrol(struct snd_sof_widget *swid
 	return 0;
 }
 
+static const struct sof_ipc4_audio_format
+*sof_ipc4_get_sink_pin_fmt(struct snd_sof_widget *swidget, int pin_id)
+{
+	struct sof_ipc4_process *process = swidget->private;
+	struct sof_ipc4_control_data *control_data = process->cdata;
+	struct sof_ipc4_base_module_cfg_ext *base_ext;
+	const struct sof_ipc4_pin_format *pin_fmt;
+	u32 module_lib_id;
+	int i;
+
+	module_lib_id = sof_ipc4_get_module_lib_id(swidget->module_info);
+	/* module_lib_id == 0 if module is in basefw */
+	if (module_lib_id == 0) {
```

Comment:
3 is not needed, process->cdata is set only if it is an init blob type

Reply:
Note that process module is the module of the snd_soc_dapm_effect type, it could be a built-in module or an external module. 
It is not likely that we will have an external module that is not snd_soc_dapm_effect type now or in the future, because by design the feature of loading external library is mainly to let the vendor not reveal their algorithm, those modules of course are the snd_soc_dapm_effect type.

This function currently is only called by process module ops, if there is external module with other asoc widget type also call this function in its ops, we will need to test on the asoc widget type to make sure swidget->private is process module.  I think this can be done in the future patches when we see this requirement, and probably this won't happen IMHO.